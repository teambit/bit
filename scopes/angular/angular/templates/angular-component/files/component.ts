import { ComponentContext } from '@teambit/generator';

export const componentFile = (context: ComponentContext) => {
  const { name, namePascalCase: Name } = context;
  return {
    relativePath: `${name}.component.ts`,
    content: `import { Component, OnInit } from '@angular/core';

@Component({
  selector: '${name}',
  template: \`
      <p>
      ${name} works!
      </p>
        \`,
  styles: [
  ]
})
export class ${Name}Component implements OnInit {
  constructor() {}

  ngOnInit(): void {}
}
`,
  };
};
