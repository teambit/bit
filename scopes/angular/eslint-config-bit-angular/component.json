{
  "componentId": {
    "name": "eslint-config-bit-angular",
    "scope": "teambit.angular"
  },
  "propagate": true,
  "extensions": {
    "teambit.pkg/pkg": {
      "packageJson": {
        "name": "@teambit/{name}",
        "private": false
      },
      "packageManagerPublishArgs": [
        "--access public"
      ]
    },
    "teambit.dependencies/dependency-resolver": {
      "policy": {
        "dependencies": {
          "@teambit/legacy": "-",
          "eslint": "-"
        },
        "devDependencies": {
          "@angular-eslint/builder": "4.0.0",
          "@angular-eslint/eslint-plugin": "4.0.0",
          "@angular-eslint/eslint-plugin-template": "4.0.0",
          "@angular-eslint/template-parser": "4.0.0",
          "@typescript-eslint/eslint-plugin": "4.16.1",
          "@typescript-eslint/parser": "4.16.1",
          "@teambit/legacy": "-",
          "eslint": "-"
        },
        "peerDependencies": {
          "eslint": "> 7.0.0",
          "@teambit/legacy": "1.0.62"
        }
      }
    }
  }
}
