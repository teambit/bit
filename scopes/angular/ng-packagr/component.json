{
  "componentId": {
    "name": "ng-packagr",
    "version": "0.0.1"
  },
  "propagate": true,
  "extensions": {
    "teambit.pkg/pkg": {
      "packageJson": {
        "name": "@teambit/{name}",
        "private": false
      },
      "packageManagerPublishArgs": [
        "--access public"
      ]
    },
    "teambit.dependencies/dependency-resolver": {
      "policy": {
        "devDependencies": {
          "@angular/compiler": "^11.0.0",
          "@angular/compiler-cli": "^11.0.0",
          "ng-packagr": "^11.0.0"
        }
      }
    }
  }
}
