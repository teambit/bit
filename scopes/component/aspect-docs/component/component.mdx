A [Bit component](https://bit.dev/teambit/component/component) is a super-set of a node package. In addition to its distributable code, a Bit component has everything it needs to be maintained and developed as an independent building block.

That includes its source code, version history, development environment configurations and documentation.

A Bit component is authored in a Bit workspace. It is where its configurations are set, where it is rendered in isolation and where it is being compiled and built.

When a component gets 'tagged', its release version gets stored in the workspace local scope as a completely isolated unit.
That means, all its configurations and built artifacts are encapsulated as a single, immutable and transferable object.
This object can then be exported and stored in a remote scope, hosted on a Bit server like [Bit.dev](https://bit.dev), where it is made available to other web projects, both as a Bit component but also as a node package.

Bit components 'imported' into a Bit workspace, can be modified, versioned and exported again.

## A Bit component common structure

Below is an example of a Bit component file structure. The types of files are determined by the environment in use although there are a few constants to be aware of:

- A documentation file (`*.docs.*`). That is a combination of an auto-generated and manually customized documentation for the component.
- 'Compositions' files (`*.compositions.*`). These are examples of that component - relevant only to UI components.
- A "barrel" or entry file (`index.ts`).

```bash
├── account/login-form           # root directory for storing all component files
    ├── index.tsx                # barrel/entry file to export modules
    ├── login-form.composite.tsx # examples of that component being used
    ├── login-form.docs.mdx      # documentation
    ├── login-form.spec.tsx      # tests
    └── login-form.tsx           # implementation
```

> All _internal dependencies_ of a component must be in the same directory (its styling, documentation, compositions, etc.)

## Stages in a Bit component's life

A Bit component goes through a number of stages before it can be used across repositories:

### 1. Authored & tracked in a Bit workspace

- A component is created and tracked by Bit.
- A package for that component is created in the workspace' `node_modules` directory (at this point this package can only be used internally).
- Configurations are set for the component using the `workspace.json` configuration file. That includes its development environment and dependency policies.
- A component is rendered in an isolated "controlled environment". These is done by authoring and rendering component 'compositions', which are instances of that component as it is being used in different context and variations.
- A component is documented (in addition to the auto-generated documentation created by its environment)

### 2. Versioned

A component gets a new release version. That happens only after it is built using its environment's build pipeline. The new release version can be determined locally or as part of a remote CI (after being 'suggested for new version release' by the local workspace).

### 3. Exported

A component's release version, encapsulated with all its configurations and built artifacts, gets uploaded to a Remote Scope (e.g, [Bit.dev](https://bit.dev)).

This exported version is now available to be used and maintained in other Bit Workspaces.

## Component Id

Each Bit component has a unique identifier with the following pattern:<br />
`<scope-owner>.<scope-name>/<namespace(s)>/<name>`. <br />
A component ID is generated when a component gets tracked by Bit for the first time.

> Note that not all Bit servers will have a 'scope-owner'

- **Scope** - The component's scope as applied by the `workspace.json` file. It can be a `scope` property as defined for the component's `variant` or the `defaultScope` configured to the `teambit.workspace/workspace` extension. `scope` is usually a combination of the scope owner and scope name (e.g, `<my-org><my-scope>`)

- **Namespaces** (optional) - Set with the `--namespace` or `-n` flag when adding the component (supports nesting - `--namespace nesting/namespace/yay`).
- **Name** - The name of the component, according to the component's root directory name.

Bit uses these IDs when listing or running operations and commands on components.

## Component package

Bit creates a package from each component in the workspace' root `node_modules` directory. This package contains the component's transpiled code for other components to import.
The package name is defined by the component ID. However, Node supports a single forward slash (`/`) in a module name (to set the module scope). Bit uses the `<scope>` defined for the components as such (with the `/` separator). All other `/` between namespaces (if found) are translated to dots (`.`).

For example the component ID `my-org.design-system/component/base/button` will result in a module called `@my-org/design-system.components.base.button`.

> Why doesn't Bit keep the same module naming as node?
>
> The future of JavaScript development brings tools like ESM and Deno that have a different approach for module consumption with namespacing support. As Bit components needs to be future-proof to support additional tooling it keeps its own naming scheme that can be translated according to the consumption method.

## Component configurations

A component's configurations are set by the `workspace.jsonc` configuration file. These settings include rules and policies regarding its dependencies, the development environment to support it, the properties for its generated package, and so on.

A component can also have its configurations "ejected". That means it will stop inheriting its settings from the `workspace.jsonc` file. Instead, these will be managed using the `component.json` configuration file, located in its directory (this file will only appear after ejecting).

To eject a component, run the following:

```bash
bit eject-conf <component-id>
```

> Ejecting a component's configurations is not recommended! Use the `workspace.jsonc` instead, whenever possible.
