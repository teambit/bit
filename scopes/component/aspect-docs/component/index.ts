export { default as Component } from './component.mdx';
