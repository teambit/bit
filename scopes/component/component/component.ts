import { AnyFS } from '@teambit/any-fs';
import { capitalize } from '@teambit/string.capitalize';
import { SemVer } from 'semver';
import { ComponentID } from '@teambit/component-id';
import { BitError } from '@teambit/bit-error';

import { ComponentFactory } from './component-factory';
import ComponentFS from './component-fs';
// import { NothingToSnap } from './exceptions';
import ComponentConfig from './config';
// eslint-disable-next-line import/no-cycle
import { Snap } from './snap';
import { State } from './state';
import { TagMap } from './tag-map';
import { Tag } from './tag';
// import { Author } from './types';

type SnapsIterableOpts = {
  firstParentOnly?: boolean;
  stopFn?: (snap: Snap) => Promise<boolean>;
};

/**
 * in-memory representation of a component.
 */
export class Component {
  constructor(
    /**
     * component ID represented by the `ComponentId` type.
     */
    readonly id: ComponentID,

    /**
     * head version of the component. can be `null` for new components.
     */
    readonly head: Snap | null = null,

    /**
     * state of the component.
     */
    private _state: State,

    /**
     * tags of the component.
     */
    readonly tags: TagMap = new TagMap(),

    /**
     * the component factory
     */
    private factory: ComponentFactory
  ) {}

  get state(): State {
    return this._state;
  }

  set state(state: State) {
    this._state = state;
  }

  /**
   * component configuration which is later generated to a component `package.json` and `bit.json`.
   */
  get config(): ComponentConfig {
    return this.state.config;
  }

  /**
   * in-memory representation of the component current filesystem.
   */
  get filesystem(): ComponentFS {
    return this.state.filesystem;
  }

  get headTag() {
    if (!this.head) return undefined;
    return this.tags.byHash(this.head.hash);
  }

  get latest(): string | undefined {
    if (!this.head) return undefined;
    return this.tags.getLatest();
  }

  stringify(): string {
    return JSON.stringify({
      id: this.id,
      head: this.head,
    });
  }

  /**
   * record component changes in the `Scope`.
   */
  // snap(author: Author, message = '') {
  // if (!this.isModified()) throw new NothingToSnap();
  // const snap = new Snap(this, author, message);

  // return new Component(this.id, snap, snap.state);
  // }

  /**
   * display name of the component.
   */
  get displayName() {
    const tokens = this.id.name.split('-').map((token) => capitalize(token));
    return tokens.join(' ');
  }

  /**
   * tag a component `Snap` with a semantic version. we follow SemVer specs as defined [here](https://semver.org/)).
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  tag(version: SemVer) {
    // const snap = this.snap();
    // const tag = new Tag(version, snap);
    // this.tags.set(tag);
  }

  /**
   * determines whether this component is modified in the workspace.
   */
  isModified(): Promise<boolean> {
    if (!this.head) return Promise.resolve(true);
    return Promise.resolve(this.state.isModified);
    // return Promise.resolve(this.state.hash !== this.head.hash);
  }

  /**
   * determines whether this component is new.
   */
  isNew(): Promise<boolean> {
    return Promise.resolve(this.head === null);
  }

  // TODO: @david after snap we need to make sure to refactor here.
  loadState(snapId: string): Promise<State> {
    return this.factory.getState(this.id, snapId);
  }

  loadSnap(snapId?: string): Promise<Snap> {
    const snapToGet = snapId || this.head?.hash;
    if (!snapToGet) {
      throw new BitError('could not load snap for new components');
    }
    return this.factory.getSnap(this.id, snapToGet);
  }

  /**
   * Get iterable which iterate over snap parents lazily
   * @param snapId
   * @param options
   */
  snapsIterable(snapId?: string, options: SnapsIterableOpts = {}): AsyncIterable<Snap> {
    const snapToStart = snapId || this.head?.hash;
    let nextSnaps = [snapToStart];
    let done;

    const iterator: AsyncIterator<Snap> = {
      next: async () => {
        if (done) {
          return { value: undefined, done };
        }
        const currSnapId = nextSnaps.shift();
        const snap = await this.loadSnap(currSnapId);
        if (snap.parents && snap.parents.length) {
          if (options.firstParentOnly) {
            nextSnaps.push(snap.parents[0]);
          } else {
            nextSnaps = nextSnaps.concat(snap.parents);
          }
        }
        if (!nextSnaps.length) {
          done = true;
        }
        if (options.stopFn) {
          done = await options.stopFn(snap);
        }
        return { value: snap, done: undefined };
      },
    };
    return {
      [Symbol.asyncIterator]: () => iterator,
    };
  }

  /**
   * traverse recursively from the provided snap (or head) upwards until it finds a tag
   * @param snapToStartFrom
   */
  async getClosestTag(snapToStartFrom?: string): Promise<Tag | undefined> {
    const tagsHashMap = this.tags.getHashMap();
    const stopFn = async (snap: Snap) => {
      if (tagsHashMap.has(snap.hash)) {
        return true;
      }
      return false;
    };
    const iterable = this.snapsIterable(snapToStartFrom, { firstParentOnly: true, stopFn });
    const snaps: Snap[] = [];
    for await (const snap of iterable) {
      snaps.push(snap);
    }
    if (snaps.length) {
      const hashOfLastSnap = snaps[snaps.length - 1].hash;
      return tagsHashMap.get(hashOfLastSnap);
    }
    return undefined;
  }

  /**
   * checkout the component to a different version in its working tree.
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  checkout(version: SemVer) {}

  /**
   * examine difference between two components.
   */
  // diff(other: Component): Difference {}

  /**
   * merge two different components
   */
  // merge(other: Component): Component {}

  /**
   * write a component to a given file system.
   * @param path root path to write the component
   * @param fs instance of any fs to use.
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  write(path: string, fs?: AnyFS) {}

  /**
   *
   * Check if 2 components are equal
   * @param {Component} component
   * @returns {boolean}
   * @memberof Component
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  equals(component: Component): boolean {
    return component.id.toString() === this.id.toString();
  }
}
