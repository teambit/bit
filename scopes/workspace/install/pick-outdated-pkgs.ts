import colorizeSemverDiff from '@pnpm/colorize-semver-diff';
import semverDiff from '@pnpm/semver-diff';
import { OutdatedPkg } from '@teambit/dependency-resolver';
import { omit } from 'lodash';
import { getBorderCharacters, table } from 'table';
import chalk from 'chalk';
import { prompt } from 'enquirer';
import semver from 'semver';

/**
 * Lets the user pick the packages that should be updated.
 */
export async function pickOutdatedPkgs(outdatedPkgs: OutdatedPkg[]): Promise<MergedOutdatedPkg[]> {
  const { updateDependencies } = (await prompt({
    choices: makeOutdatedPkgChoices(outdatedPkgs),
    footer: '\nEnter to start updating. Ctrl-c to cancel.',
    indicator: (state: any, choice: any) => ` ${choice.enabled ? '●' : '○'}`,
    message:
      'Choose which packages to update ' +
      `(Press ${chalk.cyan('<space>')} to select, ` +
      `${chalk.cyan('<a>')} to toggle all, ` +
      `${chalk.cyan('<i>')} to invert selection)
${chalk.green('Green')} - indicates a semantically safe update
${chalk.red('Red')} - indicates a semantically breaking change`,
    name: 'updateDependencies',
    pointer: '❯',
    styles: {
      dark: chalk.white,
      em: chalk.bgBlack.whiteBright,
      success: chalk.white,
    },
    type: 'multiselect',
    validate(value: string[]) {
      if (value.length === 0) {
        return 'You must choose at least one package.';
      }
      return true;
    },
    j() {
      return this.down();
    },
    k() {
      return this.up();
    },
    result(names: string[]) {
      // This is needed in order to have the values of the choices in the answer object.
      // Otherwise, only the names of the selected choices would've been included.
      return this.map(names);
    },
    cancel() {
      // By default, canceling the prompt via Ctrl+c throws an empty string.
      // The custom cancel function prevents that behavior.
      // Otherwise, Bit CLI would print an error and confuse users.
      // See related issue: https://github.com/enquirer/enquirer/issues/225
    },
  } as any)) as { updateDependencies: Record<string, string | MergedOutdatedPkg> };
  return Object.values(updateDependencies ?? {}).filter(
    (updateDependency) => typeof updateDependency !== 'string'
  ) as OutdatedPkg[];
}

const DEP_TYPE_PRIORITY = {
  dependencies: 0,
  devDependencies: 1,
  peerDependencies: 2,
};

/**
 * Groups the outdated packages and makes choices for enquirer's prompt.
 */
export function makeOutdatedPkgChoices(outdatedPkgs: OutdatedPkg[]) {
  const mergedOutdatedPkgs = mergeOutdatedPkgs(outdatedPkgs);
  mergedOutdatedPkgs.sort((pkg1, pkg2) => {
    if (pkg1.targetField === pkg2.targetField) return pkg1.name.localeCompare(pkg2.name);
    return DEP_TYPE_PRIORITY[pkg1.targetField] - DEP_TYPE_PRIORITY[pkg2.targetField];
  });
  const renderedTable = alignColumns(outdatedPkgsRows(mergedOutdatedPkgs));
  const groupedChoices = {};
  mergedOutdatedPkgs.forEach((outdatedPkg, index) => {
    const context = renderContext(outdatedPkg);
    if (!groupedChoices[context]) {
      groupedChoices[context] = [];
    }
    groupedChoices[context].push({
      message: renderedTable[index],
      name: outdatedPkg.name,
      value: outdatedPkg,
    });
  });
  const choices = Object.entries(groupedChoices).map(([context, subChoices]) => ({
    message: chalk.cyan(context),
    choices: subChoices,
  }));
  return choices;
}

export interface MergedOutdatedPkg extends OutdatedPkg {
  dependentComponents?: string[];
  hasDifferentRanges?: boolean;
}

function mergeOutdatedPkgs(outdatedPkgs: OutdatedPkg[]): MergedOutdatedPkg[] {
  const mergedOutdatedPkgs: Record<
    string,
    MergedOutdatedPkg & Required<Pick<MergedOutdatedPkg, 'dependentComponents'>>
  > = {};
  const outdatedPkgsNotFromComponentModel: OutdatedPkg[] = [];
  for (const outdatedPkg of outdatedPkgs) {
    if (outdatedPkg.source === 'component-model' && outdatedPkg.componentId) {
      if (!mergedOutdatedPkgs[outdatedPkg.name]) {
        mergedOutdatedPkgs[outdatedPkg.name] = {
          ...omit(outdatedPkg, ['componentId']),
          source: 'rootPolicy',
          dependentComponents: [outdatedPkg.componentId],
        };
      } else {
        if (mergedOutdatedPkgs[outdatedPkg.name].currentRange !== outdatedPkg.currentRange) {
          mergedOutdatedPkgs[outdatedPkg.name].hasDifferentRanges = true;
        }
        mergedOutdatedPkgs[outdatedPkg.name].currentRange = tryPickLowestRange(
          mergedOutdatedPkgs[outdatedPkg.name].currentRange,
          outdatedPkg.currentRange
        );
        mergedOutdatedPkgs[outdatedPkg.name].dependentComponents.push(outdatedPkg.componentId);
        if (outdatedPkg.targetField === 'dependencies') {
          mergedOutdatedPkgs[outdatedPkg.name].targetField = outdatedPkg.targetField;
        }
      }
    } else {
      outdatedPkgsNotFromComponentModel.push(outdatedPkg);
    }
  }
  return [...Object.values(mergedOutdatedPkgs), ...outdatedPkgsNotFromComponentModel];
}

function tryPickLowestRange(range1: string, range2: string) {
  if (range1 === '*' || range2 === '*') return '*';
  try {
    return semver.lt(rangeToVersion(range1), rangeToVersion(range2)) ? range1 : range2;
  } catch {
    return '*';
  }
}

function rangeToVersion(range: string) {
  if (range.startsWith('~') || range.startsWith('^')) {
    return range.substring(1);
  }
  return range;
}

function renderContext(outdatedPkg: MergedOutdatedPkg) {
  if (outdatedPkg.variantPattern) {
    return `${outdatedPkg.variantPattern} (variant)`;
  }
  if (outdatedPkg.componentId) {
    return `${outdatedPkg.componentId} (component)`;
  }
  return 'Root policies';
}

const TARGET_FIELD_TO_DEP_TYPE = {
  devDependencies: 'dev',
  dependencies: 'runtime',
  peerDependencies: 'peer',
};

function outdatedPkgsRows(outdatedPkgs: MergedOutdatedPkg[]) {
  return outdatedPkgs.map((outdatedPkg) => {
    const { change, diff } = semverDiff(outdatedPkg.currentRange, outdatedPkg.latestRange);
    let colorizeChange = change ?? 'breaking';
    if (change === 'feature') {
      colorizeChange = 'fix';
    }
    const latest = colorizeSemverDiff({
      change: colorizeChange,
      diff,
    });
    return [
      outdatedPkg.name,
      chalk.grey(`(${TARGET_FIELD_TO_DEP_TYPE[outdatedPkg.targetField]})`),
      outdatedPkg.hasDifferentRanges ? '*' : outdatedPkg.currentRange,
      '❯',
      latest,
      outdatedPkg.dependentComponents ? renderDependents(outdatedPkg.dependentComponents) : ' ',
    ];
  });
}

function renderDependents(dependentComponents: string[]): string {
  let result = `because of ${dependentComponents[0]}`;
  if (dependentComponents.length > 1) {
    result += ` and ${dependentComponents.length - 1} other components`;
  }
  return result;
}

function alignColumns(rows: string[][]) {
  return table(rows, {
    border: getBorderCharacters('void'),
    columnDefault: {
      paddingLeft: 0,
      paddingRight: 1,
    },
    columns: {
      // This is the current range column
      2: { alignment: 'right' },
    },
    drawHorizontalLine: () => false,
  }).split('\n');
}
