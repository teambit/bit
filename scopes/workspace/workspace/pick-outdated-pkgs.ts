import colorizeSemverDiff from '@pnpm/colorize-semver-diff';
import semverDiff from '@pnpm/semver-diff';
import { OutdatedPkg } from '@teambit/dependency-resolver';
import { getBorderCharacters, table } from 'table';
import chalk from 'chalk';
import { prompt } from 'enquirer';

/**
 * Lets the user pick the packages that should be updated.
 */
export async function pickOutdatedPkgs(outdatedPkgs: OutdatedPkg[]): Promise<OutdatedPkg[]> {
  const { updateDependencies } = (await prompt({
    choices: makeOutdatedPkgChoices(outdatedPkgs),
    footer: '\nEnter to start updating. Ctrl-c to cancel.',
    indicator: (state: any, choice: any) => ` ${choice.enabled ? '●' : '○'}`,
    message:
      'Choose which packages to update ' +
      `(Press ${chalk.cyan('<space>')} to select, ` +
      `${chalk.cyan('<a>')} to toggle all, ` +
      `${chalk.cyan('<i>')} to invert selection)
${chalk.green('Green')} - indicates a semantically safe update
${chalk.red('Red')} - indicates a semantically breaking change`,
    name: 'updateDependencies',
    pointer: '❯',
    styles: {
      dark: chalk.white,
      em: chalk.bgBlack.whiteBright,
      success: chalk.white,
    },
    type: 'multiselect',
    validate(value: string[]) {
      if (value.length === 0) {
        return 'You must choose at least one package.';
      }
      return true;
    },
    j() {
      return this.down();
    },
    k() {
      return this.up();
    },
    result(names: string[]) {
      // This is needed in order to have the values of the choices in the answer object.
      // Otherwise, only the names of the selected choices would've been included.
      return this.map(names);
    },
    cancel() {
      // By default, canceling the prompt via Ctrl+c throws an empty string.
      // The custom cancel function prevents that behavior.
      // Otherwise, Bit CLI would print an error and confuse users.
      // See related issue: https://github.com/enquirer/enquirer/issues/225
    },
  } as any)) as { updateDependencies: Record<string, string | OutdatedPkg> };
  return Object.values(updateDependencies ?? {}).filter(
    (updateDependency) => typeof updateDependency !== 'string'
  ) as OutdatedPkg[];
}

const DEP_TYPE_PRIORITY = {
  dependencies: 0,
  devDependencies: 1,
  peerDependencies: 2,
};

/**
 * Groups the outdated packages and makes choices for enquirer's prompt.
 */
export function makeOutdatedPkgChoices(outdatedPkgs: OutdatedPkg[]) {
  outdatedPkgs.sort((pkg1, pkg2) => {
    if (pkg1.targetField === pkg2.targetField) return pkg1.name.localeCompare(pkg2.name);
    return DEP_TYPE_PRIORITY[pkg1.targetField] - DEP_TYPE_PRIORITY[pkg2.targetField];
  });
  const renderedTable = alignColumns(outdatedPkgsRows(outdatedPkgs));
  const groupedChoices = {};
  outdatedPkgs.forEach((outdatedPkg, index) => {
    const context = renderContext(outdatedPkg);
    if (!groupedChoices[context]) {
      groupedChoices[context] = [];
    }
    groupedChoices[context].push({
      message: renderedTable[index],
      name: outdatedPkg.name,
      value: outdatedPkg,
    });
  });
  const choices = Object.entries(groupedChoices).map(([context, subChoices]) => ({
    message: chalk.cyan(context),
    choices: subChoices,
  }));
  return choices;
}

function renderContext(outdatedPkg: OutdatedPkg) {
  if (outdatedPkg.variantPattern) {
    return `${outdatedPkg.variantPattern} (variant)`;
  }
  if (outdatedPkg.componentId) {
    return `${outdatedPkg.componentId} (component)`;
  }
  return 'Root policies';
}

const TARGET_FIELD_TO_DEP_TYPE = {
  devDependencies: 'dev',
  dependencies: 'runtime',
  peerDependencies: 'peer',
};

function outdatedPkgsRows(outdatedPkgs: OutdatedPkg[]) {
  return outdatedPkgs.map((outdatedPkg) => {
    const { change, diff } = semverDiff(outdatedPkg.currentRange, outdatedPkg.latestRange);
    let colorizeChange = change ?? 'breaking';
    if (change === 'feature') {
      colorizeChange = 'fix';
    }
    const latest = colorizeSemverDiff({
      change: colorizeChange,
      diff,
    });
    return [
      outdatedPkg.name,
      chalk.grey(`(${TARGET_FIELD_TO_DEP_TYPE[outdatedPkg.targetField]})`),
      outdatedPkg.currentRange,
      '❯',
      latest,
    ];
  });
}

function alignColumns(rows: string[][]) {
  return table(rows, {
    border: getBorderCharacters('void'),
    columnDefault: {
      paddingLeft: 0,
      paddingRight: 1,
    },
    columns: {
      // This is the current range column
      2: { alignment: 'right' },
    },
    drawHorizontalLine: () => false,
  }).split('\n');
}
