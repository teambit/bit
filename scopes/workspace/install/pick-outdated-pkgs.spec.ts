import { ComponentID } from '@teambit/component';
import { makeOutdatedPkgChoices } from './pick-outdated-pkgs';

describe('makeOutdatedPkgChoices', () => {
  it('should render choices in correct order', () => {
    const choices = makeOutdatedPkgChoices([
      {
        name: 'zoo',
        currentRange: '1.0.0',
        latestRange: '1.1.0',
        source: 'rootPolicy',
        targetField: 'devDependencies',
      },
      {
        name: 'qar',
        currentRange: '1.0.0',
        latestRange: '1.1.0',
        source: 'rootPolicy',
        targetField: 'dependencies',
      },
      {
        name: 'foo',
        currentRange: '1.0.0',
        latestRange: '2.0.0',
        source: 'rootPolicy',
        targetField: 'dependencies',
      },
      {
        name: 'bar',
        currentRange: '1.0.0',
        latestRange: '1.1.0',
        source: 'rootPolicy',
        targetField: 'peerDependencies',
      },
    ]);
    // @ts-ignore
    expect(choices).toMatchSnapshot();
  });
  it('should render choices with context information', () => {
    const choices = makeOutdatedPkgChoices([
      {
        name: 'foo',
        currentRange: '1.0.0',
        latestRange: '2.0.0',
        source: 'component',
        componentId: ComponentID.fromString('scope/comp1'),
        targetField: 'dependencies',
      },
      {
        name: 'bar',
        currentRange: '1.0.0',
        latestRange: '1.1.0',
        source: 'variants',
        variantPattern: '{comp2}',
        targetField: 'peerDependencies',
      },
    ]);
    // @ts-ignore
    expect(choices).toMatchSnapshot();
  });
  it('should group component model updates of the same dependency', () => {
    const choices = makeOutdatedPkgChoices([
      {
        name: 'foo',
        currentRange: '1.0.0',
        latestRange: '2.0.0',
        source: 'component-model',
        componentId: ComponentID.fromString('scope/comp1'),
        targetField: 'devDependencies',
      },
      {
        name: 'foo',
        currentRange: '1.1.0',
        latestRange: '2.0.0',
        source: 'component-model',
        componentId: ComponentID.fromString('scope/comp2'),
        targetField: 'dependencies',
      },
    ]);
    // @ts-ignore
    expect(choices).toMatchSnapshot();
  });
  it("should group component model updates of the same dependency and use * as current range when can't compare ranges", () => {
    const choices = makeOutdatedPkgChoices([
      {
        name: 'foo',
        currentRange: '<=10.0.0',
        latestRange: '2.0.0',
        source: 'component-model',
        componentId: ComponentID.fromString('scope/comp1'),
        targetField: 'dependencies',
      },
      {
        name: 'foo',
        currentRange: '1.1.0',
        latestRange: '2.0.0',
        source: 'component-model',
        componentId: ComponentID.fromString('scope/comp2'),
        targetField: 'dependencies',
      },
    ]);
    // @ts-ignore
    expect(choices).toMatchSnapshot();
  });
  it('should group component model updates of the same dependency and display the current range when all components use the same range', () => {
    const choices = makeOutdatedPkgChoices([
      {
        name: 'foo',
        currentRange: '^1.2.3',
        latestRange: '2.0.0',
        source: 'component-model',
        componentId: ComponentID.fromString('scope/comp1'),
        targetField: 'dependencies',
      },
      {
        name: 'foo',
        currentRange: '^1.2.3',
        latestRange: '2.0.0',
        source: 'component-model',
        componentId: ComponentID.fromString('scope/comp2'),
        targetField: 'dependencies',
      },
    ]);
    // @ts-ignore
    expect(choices).toMatchSnapshot();
  });
});
