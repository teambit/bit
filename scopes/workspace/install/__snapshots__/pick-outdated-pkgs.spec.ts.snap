// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeOutdatedPkgChoices should group component model updates of the same dependency 1`] = `
Array [
  Object {
    "choices": Array [
      Object {
        "message": "foo [90m(runtime)[39m * ‚ùØ [91m[1m2.0.0[22m[39m because of comp1 and 1 other components ",
        "name": "foo",
        "value": Object {
          "currentRange": "1.0.0",
          "dependentComponents": Array [
            "comp1",
            "comp2",
          ],
          "hasDifferentRanges": true,
          "latestRange": "2.0.0",
          "name": "foo",
          "source": "rootPolicy",
          "targetField": "dependencies",
        },
      },
    ],
    "message": "[36mRoot policies[39m",
  },
]
`;

exports[`makeOutdatedPkgChoices should group component model updates of the same dependency and display the current range when all components use the same range 1`] = `
Array [
  Object {
    "choices": Array [
      Object {
        "message": "foo [90m(runtime)[39m ^1.2.3 ‚ùØ [91m[1m2.0.0[22m[39m because of comp1 and 1 other components ",
        "name": "foo",
        "value": Object {
          "currentRange": "^1.2.3",
          "dependentComponents": Array [
            "comp1",
            "comp2",
          ],
          "latestRange": "2.0.0",
          "name": "foo",
          "source": "rootPolicy",
          "targetField": "dependencies",
        },
      },
    ],
    "message": "[36mRoot policies[39m",
  },
]
`;

exports[`makeOutdatedPkgChoices should group component model updates of the same dependency and use * as current range when can't compare ranges 1`] = `
Array [
  Object {
    "choices": Array [
      Object {
        "message": "foo [90m(runtime)[39m * ‚ùØ [91m[1m2.0.0[22m[39m because of comp1 and 1 other components ",
        "name": "foo",
        "value": Object {
          "currentRange": "*",
          "dependentComponents": Array [
            "comp1",
            "comp2",
          ],
          "hasDifferentRanges": true,
          "latestRange": "2.0.0",
          "name": "foo",
          "source": "rootPolicy",
          "targetField": "dependencies",
        },
      },
    ],
    "message": "[36mRoot policies[39m",
  },
]
`;

exports[`makeOutdatedPkgChoices should render choices in correct order 1`] = `
Array [
  Object {
    "choices": Array [
      Object {
        "message": "foo [90m(runtime)[39m 1.0.0 ‚ùØ [91m[1m2.0.0[22m[39m   ",
        "name": "foo",
        "value": Object {
          "currentRange": "1.0.0",
          "latestRange": "2.0.0",
          "name": "foo",
          "source": "rootPolicy",
          "targetField": "dependencies",
        },
      },
      Object {
        "message": "qar [90m(runtime)[39m 1.0.0 ‚ùØ 1.[92m[1m1.0[22m[39m   ",
        "name": "qar",
        "value": Object {
          "currentRange": "1.0.0",
          "latestRange": "1.1.0",
          "name": "qar",
          "source": "rootPolicy",
          "targetField": "dependencies",
        },
      },
      Object {
        "message": "zoo [90m(dev)[39m     1.0.0 ‚ùØ 1.[92m[1m1.0[22m[39m   ",
        "name": "zoo",
        "value": Object {
          "currentRange": "1.0.0",
          "latestRange": "1.1.0",
          "name": "zoo",
          "source": "rootPolicy",
          "targetField": "devDependencies",
        },
      },
      Object {
        "message": "bar [90m(peer)[39m    1.0.0 ‚ùØ 1.[92m[1m1.0[22m[39m   ",
        "name": "bar",
        "value": Object {
          "currentRange": "1.0.0",
          "latestRange": "1.1.0",
          "name": "bar",
          "source": "rootPolicy",
          "targetField": "peerDependencies",
        },
      },
    ],
    "message": "[36mRoot policies[39m",
  },
]
`;

exports[`makeOutdatedPkgChoices should render choices with context information 1`] = `
Array [
  Object {
    "choices": Array [
      Object {
        "message": "foo [90m(runtime)[39m 1.0.0 ‚ùØ [91m[1m2.0.0[22m[39m   ",
        "name": "foo",
        "value": Object {
          "componentId": "comp1",
          "currentRange": "1.0.0",
          "latestRange": "2.0.0",
          "name": "foo",
          "source": "component",
          "targetField": "dependencies",
        },
      },
    ],
    "message": "[36mcomp1 (component)[39m",
  },
  Object {
    "choices": Array [
      Object {
        "message": "bar [90m(peer)[39m    1.0.0 ‚ùØ 1.[92m[1m1.0[22m[39m   ",
        "name": "bar",
        "value": Object {
          "currentRange": "1.0.0",
          "latestRange": "1.1.0",
          "name": "bar",
          "source": "variants",
          "targetField": "peerDependencies",
          "variantPattern": "{comp2}",
        },
      },
    ],
    "message": "[36m{comp2} (variant)[39m",
  },
]
`;
